### PR TITLE
Allow apptainer version

### LINKED ISSUE #111: Apptainer not supported?
Instead of installing singularity I tried apptainer which symlinks singularity executable. 

It however gave following error

<details>
<summary>SingularityVersionException: Wrong version (1.0.2) of singularity found, require version 3.6.0</summary>

```
File ~/mamba/envs/ewatercycle/lib/python3.10/site-packages/grpc4bmi/bmi_client_singularity.py:198, in BmiClientSingularity.__init__(self, image, work_dir, input_dirs, delay, timeout, capture_logs)
    195     msg = f'type of argument "input_dirs" must be collections.abc.Iterable; ' \
    196           f'got {qualified_name(input_dirs)} instead'
    197     raise TypeError(msg)
--> 198 check_singularity_version()
    199 host = 'localhost'
    200 port = BmiClient.get_unique_port(host)

File ~/mamba/envs/ewatercycle/lib/python3.10/site-packages/grpc4bmi/bmi_client_singularity.py:25, in check_singularity_version()
     23 local_version = semver.VersionInfo.parse(stdout.decode('utf-8').replace('_', '-'))
     24 if local_version < REQUIRED_SINGULARITY_VERSION:
---> 25     raise SingularityVersionException(f'Wrong version ({local_version}) of singularity found, '
     26                                       f'require version {REQUIRED_SINGULARITY_VERSION}')
     27 return True

SingularityVersionException: Wrong version (1.0.2) of singularity found, require version 3.6.0
```

</details>

### PR DESCRIPTION
**Pull request by sverhoeven**
Fixes #111 

Also clear away the bitrot:
* lower number of CI runs
* use Python 3.10 in CI
* Ignore files generated by protoc in SonarCloud
* Add test for  BmiR to raise coverage to green level
* pin protobuf see #115

### CODE REVIEW

**Review by Peter9192 — APPROVED** [review]
Looks good, few suggestions. CI is green but there's still a warning in the tests. Wanna fix that as well?

```
 =============================== warnings summary ===============================
test/test_singularity.py::test_from_notebook
  /opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/zmq/_future.py:498: DeprecationWarning: There is no current event loop
    f = future or self._Future()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform linux, python 3.10.6-final-0 -----------
Coverage XML written to file coverage.xml

================== 195 passed, 1 warning in 202.91s (0:03:22) ==================
```

File: requirements.txt [hunk 1]
```diff
@@ -13,8 +13,9 @@ scipy
 docker
 pytest
 pytest-cov
-futures
 https://github.com/eWaterCycle/bmi-python/archive/master.zip#egg=bmi-heat
+# Pin scipy as bmi-heat needs it for `from scipy import random`
+scipy==1.8.1
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply requirements.txt]: I updated our own fork (https://github.com/eWaterCycle/bmi-python) to use numpys random.

File: grpc4bmi/bmi_client_singularity.py [hunk 2]
```diff
@@ -10,21 +10,29 @@
...
+    (app, _, version) = version_output.replace('_', '-').split(' ')
+    local_version = semver.VersionInfo.parse(version)
+    if app == 'singularity' and local_version < REQUIRED_SINGULARITY_VERSION:
+        raise SingularityVersionException(f'Wrong version ({local_version}) of singularity found, '
+                                          f'require version {REQUIRED_SINGULARITY_VERSION}')
+    # Apptainer creates /usr/bin/singularity symlink, so if installed will report the apptainer version.
+    if app == 'apptainer' and local_version < REQUIRED_APPTAINER_VERSION:
+        raise ApptainerVersionException(f'Wrong version ({local_version}) of apptainer found, '
+                                          f'require version {REQUIRED_APPTAINER_VERSION}')
+    return True
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply grpc4bmi/bmi_client_singularity.py]: I switched to using [PEP 440](https://peps.python.org/pep-0440/) version spec and https://packaging.pypa.io/en/latest/ package to check if installed version is supported.

[the remark this answers is not in the document]
Comment (Peter9192) [reply grpc4bmi/bmi_client_singularity.py]: Much better, and learned something :-)

**Pull request merged**